### ISSUE TITLE
BLOCKSRV: Crash on restart with too many synoptics

### ISSUE THREAD

**Issue by John-Holt-Tessella** [ISISComputingGroup/IBEX#2824]
BLOCKSRV crashes if there are too many synoptics and the block cache is running. To reporduce:

1. Add  ~40 synoptics to the settings
1. console to block server and block cache
1. Make sure block cahce is running
1. Restart blockserver

You should see:

```
Alarm server will update in 20 seconds from now

INFO: Waiting for runcontrol IOC to start ...
INFO: ... Runcontrol IOC started
INFO: Runcontrol IOC started
INFO: Adding PV SYNOPTICS:NAMES
INFO: Adding PV SYNOPTICS:GET_DEFAULT
INFO: Adding PV SYNOPTICS:__BLANK__:GET
INFO: Adding PV SYNOPTICS:SET_DETAILS
INFO: Adding PV SYNOPTICS:DELETE
INFO: Adding PV SYNOPTICS:SCHEMA
INFO: Adding PV SYNOPTICS:SDASD:GET
INFO: Adding PV SYNOPTICS:FREDDI:GET
INFO: Adding PV SYNOPTICS:SAMPLE:GET
INFO: Adding PV SYNOPTICS:LARMOR:GET

@@@ @@@ @@@ @@@ @@@
@@@ Received a sigChild for process 16836. The process was killed by signal 11
@@@ Current time: Mon Dec 11 09:55:06 2017
@@@ Child process is shutting down, auto restart is disabled
@@@ ^R or ^X restarts the child, ^Q quits the server
@@@ Restarting child "BLOCKSVR"
@@@    (as C:\Instrument\Apps\EPICS\ISIS\inst_servers\master\start_blockserver_c
md.bat)
@@@ The PID of new child "BLOCKSVR" is: 14688
@@@ @@@ @@@ @@@ @@@
```

Also see #1770 for previous crash.

**Comment by ChrisM-S** [comment]
The dump above has the line 

`@@@ Child process is shutting down, auto restart is disabled`

Would enabling a restart not be a good option for preserving beam time - I'm guessing failure of the instrument/scripts/logging is guaranteed if it does not restart - in quiet hours this could be quite significant.

Might a script survive as long as it didn't call a block during the loss of the block-server or after the restart?

**Comment by AdrianPotter** [comment]
My guess is John did that manually. The Blockserver does automatically restart in general

**Comment by John-Holt-Tessella** [comment]
Yes, I switched of auto restart during the test to make it easier for me.

**Comment by Tom-Willemsen** [comment]
Only happens if you have a client open. You can make the problem increasingly consistent by having multiple clients open.

**Comment by Tom-Willemsen** [comment]
This problem has been traced back to a null pointer dereference in an EPICS DLL. I ran an older version of pcaspy (0.6.3) with debug symbols and got the following screenshot. Note the line numbers may not be correct as I don't have quite the right source files.

![image](https://user-images.githubusercontent.com/24375813/33983327-b563b606-e0ab-11e7-8d8e-05da1cda0b62.png)

@FreddieAkeroyd is going to help me build a version of pcaspy 0.7.0 with debug symbols so that we can post this problem to the tech-talk mailing list. Building pcaspy requires visual studio 2008 as it is a python extension. Other than posting the problem to tech-talk I don't think there is much we can do about this issue.

**Comment by Tom-Willemsen** [comment]
Stack trace with pcaspy 0.7.0 built against EPICS base 3.15:

![image](https://user-images.githubusercontent.com/24375813/34097794-298ca776-e3d2-11e7-9438-2f2aeea0c2e0.png)

Probably best to discuss with @FreddieAkeroyd to see if there's anything obviously wrong here (I don't have a detailed enough understanding of the internals of channel access), otherwise will post it to the tech-talk list. Mutex having an ID of NULL seems surprising but maybe that is normal?

**Comment by FreddieAkeroyd** [comment]
this->pPV being NULL is not good either, as it must have been non NULL at the start of the function as this is tested for by the if statement. Given both the mutex and pPV are NULL, it would seem likely that the casPVI instance has either been deleted or overwritten. Its destructor does not zero pPV directly, I see that casPVDestroyNotify() does zero pPV and can indirectly lead to the destructor of casPVI being called and so both get to be NULL, however casPVDestroyNotify takes out a lock and so should not run if ::read is in progress!

**Comment by FreddieAkeroyd** [comment]
OK, if the destroy event is raised by the same thread doing the read() then taking the lock doesn't help as it is already held, maybe that is what is going on. Probably need to turn on some lock debugging and catch the call to casPVDestroyNotify() - I expect if you tried putting a breakpoint there you would get lots of calls that are fine, so we probably need to trap a call to lock() in casPVDestroyNotify() when the lock is already held by the same thread.  Compiling episc mutex with  LOG_LAST_OWNER and then checking the field it sets will probably help.

**Comment by FreddieAkeroyd** [comment]
It would seem the above comment is what is going on. On compiling fully in debug mode it looks like the mutex id is non-NULL but pPV is NULL. I modified epicsMutex() to include a check for recursive calls, and then got it to raise a DebugBreak() if it was about to zero pPV with a recursion count > 1. This gave the following stack trace
![pic](https://user-images.githubusercontent.com/1132753/34653544-1b750322-f3e5-11e7-87bf-959d2e3e7e3a.png)
 
The read of a PV calls into python and a delete of the PV has perviously been triggered and gets actioned during this time in the interpreter, it looks like this all happens on the same main thread and so the lock in the C++ side does not help stop it. I tried putting in a bit of protection in the C++ side (check for pPV going NULL during a read etc) but this just pushed the error further down the line into an internal python crash. 

Looking at the python blockserver call, it seems to add PVs by passing a complete dictionary of all PVs created so far, this will re-create existing PVs and the previous references now removed from the PV dictionary will probably end up with their destructors being called on next python garbage collection. I'll post a fix to the blockserver to remove this behavior, what i'm not sure about is if there is still an underlying problem of if the blockserver deleted a PV for another reason that happened to coincide with a read of that PV from the channel access side, will need to test this separately.
